### PR TITLE
Polynomial regression surrogate added to STM together with its documentation and tests.

### PR DESCRIPTION
**Pull request by grmnptr**
A new polynomial regression surrogate has been added to STM. Documentation of the new objects, together with tests has also been provided.
(refs #15361)

### CODE REVIEW

**Comment by zachmprince** [comment]
@aeslaughter I think we need to make @grmnptr a collaborator so the tests will run automatically for him.

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+void
+PolynomialRegressionTrainer::finalize()
+{
+  gatherSum(_matrix.get_values());
+  gatherSum(_rhs.get_values());
+
+  DenseVector<Real> sol;
+  _matrix.lu_solve(_rhs, sol);
+  _coeff = sol.get_values();
```
Comment (zachmprince) [inline modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C:125]: I just wanted to leave note here for @aeslaughter and @cbolisetti in regards to what we are leaning in our MIT course. In this model we have a quadratic objective function:
```
f(a) = \sum_{i=1}^n (a_0 + a_1 x_{1,i} + a_2 x_{2,i} + ... - y_i)^2 = ||Xa - y||^2
```
So we are trying to do min_a [f(a)]. Taking the gradient, we get:
```
\nabla f(a) = X^T X a - X^T y = 0 ---> (X^T X) a = X^T y
```
Which we can solve as linear system of equations:
```
a = (X^T X)^-1X^T y
```
This is exactly what Peter has implemented here. However, we can actually solve this with gradient descent like:
```
a^(k+1) = a^(k) - eta \nabla f(a^(k))
```
With newton iteration, we actually get back the linear system:
```
a^(k+1) = a^(k) - [Hf(a^(k))]^-1 \nabla f(a^(k)) = a^(k) - (X^T X)^-1( X^T X a^(k) - X^T y) = (X^T X)^-1 X^T y
```
So my thinking is that we can create a generalized optimization class/utility that performs a version of gradient descent where the surrogate provides a function for `f(a)` and `\nabla f(a)`. `Hf(a)` could be defined explicitly by the surrogate, computed with AD, or simply left out.

Thoughts?

**Review by zachmprince — COMMENTED** [review]
More great work Peter! So far I have just looked at the source code and tests. Looks pretty good, just left a small comment regarding some micro-optimization of the training. I will need to wait until tomorrow to compile and review the documentation, put initially it looks good.

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      for (unsigned int ii = 0; ii < data.size(); ++ii)
+        i_value *= pow(data[ii], i_powers[ii]);
+
+      for (unsigned int j = 0; j < _n_poly_terms; ++j)
+      {
+        std::vector<unsigned int> j_powers(_power_matrix[j]);
+
+        Real j_value(1.0);
+        for (unsigned int jj = 0; jj < data.size(); ++jj)
+          j_value *= pow(data[jj], j_powers[jj]);
```
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C]: Good idea! I put it in the code this morning. Thanks Zach!

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+void
+PolynomialRegressionTrainer::initialize()
+{
+  // Check if results of samples matches number of samples
+  __attribute__((unused)) dof_id_type num_rows =
+      _values_distributed ? _sampler->getNumberOfLocalRows() : _sampler->getNumberOfRows();
+  mooseAssert(num_rows == _values_ptr->size(),
+              "Sampler number of rows does not match number of results from vector postprocessor.");
+}
```
Comment (zachmprince) [inline modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C:80]: Could you add a check here such that `_n_poly_terms < _sampler->getNumberofRows()`.

**Review by zachmprince — COMMENTED** [review]
Documentation looks great. It actually reminded me of one more check you should perform in the trainer.

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C (hunk 3 above)
Comment (zachmprince) [reply modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C:80]: Oops, didn't see that. All good then

**Review by zachmprince — DISMISSED** [review]

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C (hunk 1 above)
Comment (cbolisetti) [reply modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C:132]: We have an SGD implementation [here](https://github.com/idaholab/mastodon/blob/devel/src/utils/MastodonUtils.C) in `maximizeLogLikelihood` (but it needs to be its own util), that can be used as a starting point..

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C (hunk 3 above)
Comment (aeslaughter) [reply modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C:80]: Can this be promoted to a mooseError and have a test associated. This check shouldn't impact performance and having it just go on in opt mode (the default) could lead to confusing behavior.

You could also make this a `paramError("results_vpp", ...)` and also include the "results_vector" name in the message to make it very clear to the origin of the error.

**Review by aeslaughter — CHANGES_REQUESTED** [review]
https://github.com/idaholab/moose/pull/15371/files#r434113859

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C (hunk 3 above)
Comment (grmnptr) [reply modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C:87]: I implemented a paramTest with a test. Let me know if it is not enough.

**Review by zachmprince — COMMENTED** [review]
Looks really good Peter. I just had some pretty minor comments. Your analysis in the example is amazing.

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+  params.addRequiredParam<SamplerName>("sampler", "Training set defined by a sampler object.");
+  params.addRequiredParam<VectorPostprocessorName>(
+      "results_vpp", "Vectorpostprocessor with results of samples created by trainer.");
+  params.addRequiredParam<std::string>(
+      "results_vector",
+      "Name of vector from vectorpostprocessor with results of samples created by trainer");
+  params.addRequiredParam<unsigned int>("max_degree",
+                                        "Maximum polynomial degree to use for the regression.");
+  params.addParam<Real>("penalty", 0.0, "Penalty for Ridge regularization.");
```
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C:32]: The new parameter has been added together with the warning.

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  return params;
+}
+
+PolynomialRegressionTrainer::PolynomialRegressionTrainer(const InputParameters & parameters)
+  : SurrogateTrainer(parameters),
+    _coeff(declareModelData<std::vector<Real>>("_coeff")),
+    _power_matrix(declareModelData<std::vector<std::vector<unsigned int>>>("_power_matrix")),
+    _max_degree(
+        declareModelData<unsigned int>("_max_degree", getParam<unsigned int>("max_degree"))),
+    _penalty(declareModelData<Real>("_penalty", getParam<Real>("penalty")))
```
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C]: Good point! Thanks! I removed the line.

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/poly_regression_surrogate.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+!listing surrogates/polynomial_regression/normal_train.i block=MultiApps Controls Transfers VectorPostprocessors
+
+The next step is to set up two `Trainer` objects to generate the surrogate models
+from the available data. This can be done in the `Trainers` block. It is visible that
+both examples use the data from `Sampler` and `VectorPostprocessor` objects. A polynomial chaos surrogate of
+order 8 and a polynomial regression surrogate with a
+polynomial of degree at most 4 is used in this study.
+The [PolynomialChaosTrainer.md] also needs knowledge about the underlying parameter distributions
+to be able to select matching polynomials.
+
```
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/stochastic_tools/doc/content/modules/stochastic_tools/examples/poly_regression_surrogate.md:129]: Sure! I added it. I don't know how I could forget it. Thanks Zach for noticing it!

File: modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C (hunk 5 above)
[the remark this answers is not in the document]
Comment (grmnptr) [reply modules/stochastic_tools/src/surrogates/PolynomialRegressionTrainer.C]: Good point! I removed the line.

File: modules/stochastic_tools/include/utils/MultiDimPolynomialGenerator.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * Polynomials and quadratures based on defined distributions for Polynomial Chaos
+ */
+namespace StochasticTools
+{
+/**
+ Class containing functionalitties to generate the degress of the 1D Polynomials
+ within every term of a multi-dimensional polynomial.
+ */
+class MultiDimPolynomialGenerator
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/MultiDimPolynomialGenerator.h:26]: You don't need a class here, since the two methods are both static. Just make a nested namespace, the use will be the same.

**Review by aeslaughter — CHANGES_REQUESTED** [review]
This is fantastic! One more little change.

File: modules/stochastic_tools/include/utils/MultiDimPolynomialGenerator.h (hunk 7 above)
Comment (grmnptr) [reply modules/stochastic_tools/include/utils/MultiDimPolynomialGenerator.h:26]: So I was thinking a bit about this class when I wrote the code and I think it might be nice to implement a generic class for polynomials (Hermite, Lagrange, Regular etc) in the utilities. Just like the one which @zachmprince uses for the quadratures in PCE but without the quadrature options. I think then this multiD polynomial class could inherit from the 1D polynomial class and create multiD polynomials. I think in that class these functions wouldn't be static. This is the reason why I created a class instead of a namespace.  Let me know what you think. Should I still change it to a nested namespace?

Comment (zachmprince) [reply modules/stochastic_tools/include/utils/MultiDimPolynomialGenerator.h:26]: This is actually a decent idea from Peter. But I don't think this level of abstraction is needed right now. If we find another application that needs these polynomials or polynomial regression would like to use orthogonal polynomials, then we can work on making multiD polynomials more general. For now, lets just go with the namespace implementation.

Comment (aeslaughter) [reply modules/stochastic_tools/include/utils/MultiDimPolynomialGenerator.h:26]: I'll buy it.

**Review by aeslaughter — APPROVED** [review]

**Review by zachmprince — APPROVED** [review]
